File: 0189-rotate-array/0189-rotate-array.java
class Solution {
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k = k % n; // Handle k > n

        // Recursive reverse function
        reverse(nums, 0, n - 1);
        reverse(nums, 0, k - 1);
        reverse(nums, k, n - 1);
    }

    private void reverse(int[] nums, int l, int h) {
        if (l < h) {
            int temp = nums[l];
            nums[l] = nums[h];
            nums[h] = temp;
            reverse(nums, l + 1, h - 1);
        }
    }
}

File: 0189-rotate-array/0189-rotate-array.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        a=len(nums)-1
        k%=a+1
        self.reverse(nums,0,a)
        self.reverse(nums,0,k-1)
        self.reverse(nums,k,a)
        
    def reverse(self,nums,i,j):
        while i<j:
            nums[i],nums[j]=nums[j],nums[i]
            i+=1
            j-=1